third/nmh/man/dist.man
.\"
.\" %nmhwarning%
.\" $Id: dist.man,v 1.1.1.1 1999-02-07 18:14:19 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH DIST %manext1% MH.6.8 [%nmhversion%]
.SH NAME
dist \- redistribute a message to additional addresses
.SH SYNOPSIS
.in +.5i
.ti -.5i
dist
\%[+folder] \%[msg] 
\%[\-annotate] \%[\-noannotate] 
.br
\%[\-inplace] \%[\-noinplace] 
\%[\-form\ formfile] 
.br
\%[\-draftfolder\ +folder] \%[\-draftmessage\ msg]
.br
\%[\-nodraftfolder]
\%[\-editor\ editor] \%[\-noedit]
.br
\%[\-whatnowproc\ program] \%[\-nowhatnowproc]
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIDist\fR is similar to \fIforw\fR.  It prepares the specified message
for redistribution to addresses that (presumably) are not on the original
address list.

The default message form contains the following elements:

.nf
.in +.5i
.ne 10
.eo
.so %etcdir%/distcomps
.ec
.in -.5i
.fi

If a file named \*(lqdistcomps\*(rq exists in the user's nmh directory, it
will be used instead of this default form.  You may specify an alternate
forms file with the switch `\-form\ formfile'.  The form used will be
prepended to the message being resent.

If the draft already exists, \fIdist\fR will ask you as to the disposition
of the draft.  A reply of \fBquit\fR will abort \fIdist\fR, leaving the
draft intact; \fBreplace\fR will replace the existing draft with a blank
skeleton; and \fBlist\fR will display the draft.

Only those addresses in \*(lqResent\-To:\*(rq, \*(lqResent\-cc:\*(rq,
and \*(lqResent\-Bcc:\*(rq will be sent.  Also, a
\*(lqResent\-Fcc:\ folder\*(rq will be honored (see \fIsend\fR\0(1)).
Note that with \fIdist\fR, the draft should contain only
\*(lqResent\-xxx:\*(rq fields and no body.  The headers and the body of
the original message are copied to the draft when the message is sent.
Use care in constructing the headers for the redistribution.

If the `\-annotate' switch is given, the  message being distributed will
be annotated with the lines:

     Resent:\ date
     Resent:\ addrs

where each address list contains as many lines as required.  This
annotation will be done only if the message is sent directly from
\fIdist\fR.  If the message is not sent immediately from \fIdist\fR,
\*(lqcomp \-use\*(rq may be used to re\-edit and send the constructed
message, but the annotations won't take place.  Normally annotations are
done inplace in order to preserve any links to the message.  You may use
the '\-noinplace' switch to change this.

See \fIcomp\fR\0(1) for a description of the `\-editor' and `\-noedit'
switches.  Note that while in the editor, the message being resent
is available through a link named \*(lq@\*(rq (assuming the default
\fIwhatnowproc\fR\0).  In addition, the actual pathname of the message is
stored in the environment variable \fB$editalt\fR, and the pathname of
the folder containing the message is stored in the environment variable
\fB$mhfolder\fR.

The `\-draftfolder\ +folder' and `\-draftmessage\ msg' switches invoke
the \fInmh\fR draft folder facility.  This is an advanced (and highly
useful) feature.  Consult the \fImh-draft\fR(5) man page for more
information.

Upon exiting from the editor, \fIdist\fR will invoke the \fIwhatnow\fR
program.  See \fIwhatnow\fR\0(1) for a discussion of available
options.  The invocation of this program can be inhibited by using the
`\-nowhatnowproc' switch.  (In truth of fact, it is the \fIwhatnow\fR
program which starts the initial edit.  Hence, `\-nowhatnowproc' will
prevent any edit from occurring.)
.Fi
^%etcdir%/distcomps~^The standard message skeleton
^or <mh\-dir>/distcomps~^Rather than the standard skeleton
^$HOME/\&.mh\(ruprofile~^The user profile
^<mh\-dir>/draft~^The draft file
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Current\-Folder:~^To find the default current folder
.Ps
^Draft\-Folder:~^To find the default draft\-folder
.Ps
^Editor:~^To override the default editor
.Ps
^fileproc:~^Program to refile the message
.Ps
^whatnowproc:~^Program to ask the \*(lqWhat now?\*(rq questions
.Sa
comp(1), forw(1), repl(1), send(1), whatnow(1)
.De
`+folder' defaults to the current folder
.Ds
`msg' defaults to cur
.Ds
`\-noannotate'
.Ds
`\-nodraftfolder'
.Ds
`\-inplace'
.Co

If a folder is given, it will become the current folder.  The message
distributed will become the current message.
.Hi
\fIDist\fR originally used headers of the form \*(lqDistribute\-xxx:\*(rq
instead of \*(lqResent\-xxx:\*(rq.  In order to conform with the ARPA
Internet standard, RFC\-822, the \*(lqResent\-xxx:\*(rq form is now used.
\fIDist\fR will recognize \*(lqDistribute\-xxx:\*(rq type headers and
automatically convert them to \*(lqResent\-xxx:\*(rq.
.Bu
\fIDist\fR does not \fIrigorously\fR check the message being distributed
for adherence to the transport standard, but \fIpost\fR called by
\fIsend\fR does.  The \fIpost\fR program will balk (and rightly so) at
poorly formatted messages, and \fIdist\fR won't correct things for you.

If \fIwhatnowproc\fR is \fIwhatnow\fR, then \fIdist\fR uses a built\-in
\fIwhatnow\fR, it does not actually run the \fIwhatnow\fR program.
Hence, if you define your own \fIwhatnowproc\fR, don't call it
\fIwhatnow\fR since \fIdist\fR won't run it.

If your current working directory is not writable, the link named
\*(lq@\*(rq is not available.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/dist.c

/*
 * dist.c -- re-distribute a message
 *
 * $Id: dist.c,v 1.1.1.1 1999-02-07 18:14:13 danw Exp $
 */

#include <h/mh.h>
#include <fcntl.h>

static struct swit switches[] = {
#define	ANNOSW	0
    { "annotate", 0 },
#define	NANNOSW	1
    { "noannotate", 0 },
#define	DFOLDSW	2
    { "draftfolder +folder", 0 },
#define	DMSGSW	3
    { "draftmessage msg", 0 },
#define	NDFLDSW	4
    { "nodraftfolder", 0 },
#define	EDITRSW	5
    { "editor editor", 0 },
#define	NEDITSW	6
    { "noedit", 0 },
#define	FORMSW	7
    { "form formfile", 0 },
#define	INPLSW	8
    { "inplace", 0 },
#define	NINPLSW	9
    { "noinplace", 0 },
#define	WHATSW	10
    { "whatnowproc program", 0 },
#define	NWHATSW	11
    { "nowhatnowproc", 0 },
#define VERSIONSW 12
    { "version", 0 },
#define	HELPSW	13
    { "help", 4 },
#define	FILESW	14
    { "file file", -4 },	/* interface from msh */
    { NULL, 0 }
};

static struct swit aqrnl[] = {
#define	NOSW	0
    { "quit", 0 },
#define	YESW	1
    { "replace", 0 },
#define	LISTDSW	2
    { "list", 0 },
#define	REFILSW	3
    { "refile +folder", 0 },
#define NEWSW	4
    { "new", 0 },
    { NULL, 0 }
};


static struct swit aqrl[] = {
    { "quit", 0 },
    { "replace", 0 },
    { "list", 0 },
    { "refile +folder", 0 },
    { NULL, 0 }
};


int
main (int argc, char **argv)
{
    int anot = 0, inplace = 1, nedit = 0;
    int nwhat = 0, i, in, isdf = 0, out;
    char *cp, *cwd, *maildir, *msgnam, *dfolder = NULL;
    char *dmsg = NULL, *ed = NULL, *file = NULL, *folder = NULL;
    char *form = NULL, *msg = NULL, buf[BUFSIZ], drft[BUFSIZ];
    char **argp, **arguments;
    struct msgs *mp = NULL;
    struct stat st;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [msg] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case ANNOSW: 
		    anot++;
		    continue;
		case NANNOSW: 
		    anot = 0;
		    continue;

		case EDITRSW: 
		    if (!(ed = *argp++) || *ed == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nedit = 0;
		    continue;
		case NEDITSW:
		    nedit++;
		    continue;
		    
		case WHATSW: 
		    if (!(whatnowproc = *argp++) || *whatnowproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nwhat = 0;
		    continue;
		case NWHATSW: 
		    nwhat++;
		    continue;

		case FILESW: 
		    if (file)
			adios (NULL, "only one file at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    file = path (cp, TFILE);
		    continue;
		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case INPLSW: 
		    inplace++;
		    continue;
		case NINPLSW: 
		    inplace = 0;
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			adios (NULL, "only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
			    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW: 
		    if (dmsg)
			adios (NULL, "only one draft message at a time!");
		    if (!(dmsg = *argp++) || *dmsg == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    if (msg)
		adios (NULL, "only one message at a time!");
	    else
		msg = cp;
	}
    }

    cwd = getcpy (pwd ());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (file && (msg || folder))
	adios (NULL, "can't mix files and folders/msgs");

    if (form) {
	if ((in = open (etcpath (form), O_RDONLY)) == NOTOK)
	    adios (form, "unable to open form file");
    } else {
	if ((in = open (etcpath (distcomps), O_RDONLY)) == NOTOK)
	    adios (distcomps, "unable to open default components file");
	form = distcomps;
    }

try_it_again:
    strncpy (drft, m_draft (dfolder, dmsg, NOUSE, &isdf), sizeof(drft));

    /* Check if draft already exists */
    if (stat (drft, &st) != NOTOK) {
	printf ("Draft \"%s\" exists (%ld bytes).", drft, (long) st.st_size);
	for (i = LISTDSW; i != YESW;) {
	    if (!(argp = getans ("\nDisposition? ", isdf ? aqrnl : aqrl)))
		done (1);
	    switch (i = smatch (*argp, isdf ? aqrnl : aqrl)) {
		case NOSW: 
		    done (0);
		case NEWSW: 
		    dmsg = NULL;
		    goto try_it_again;
		case YESW: 
		    break;
		case LISTDSW: 
		    showfile (++argp, drft);
		    break;
		case REFILSW: 
		    if (refile (++argp, drft) == 0)
			i = YESW;
		    break;
		default: 
		    advise (NULL, "say what?");
		    break;
	    }
	}
    }
    if ((out = creat (drft, m_gmprot ())) == NOTOK)
	adios (drft, "unable to create");

    cpydata (in, out, form, drft);
    close (in);
    close (out);

    if (file) {
	/*
	 * Dist a file
	 */
	anot = 0;	/* don't want to annotate a file */
    } else {
	/*
	 * Dist a message
	 */
	if (!msg)
	    msg = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder)))
	    adios (NULL, "unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    adios (NULL, "no messages in %s", folder);

	/* parse the message range/sequence/name and set SELECTED */
	if (!m_convert (mp, msg))
	    done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	if (mp->numsel > 1)
	    adios (NULL, "only one message at a time!");
    }

    msgnam = file ? file : getcpy (m_name (mp->lowsel));
    if ((in = open (msgnam, O_RDONLY)) == NOTOK)
	adios (msgnam, "unable to open message");

    if (!file) {
	context_replace (pfolder, folder);/* update current folder  */
	seq_setcur (mp, mp->lowsel);	  /* update current message */
	seq_save (mp);			  /* synchronize sequences  */
	context_save ();		  /* save the context file  */
    }

    if (nwhat)
	done (0);
    what_now (ed, nedit, NOUSE, drft, msgnam, 1, mp,
	anot ? "Resent" : NULL, inplace, cwd);
    done (1);
}
